routing/triert/table.go

```diff
@@ -0,0 +1,227 @@
+	}
+	rt.keys.Store(&trie.Trie{})
+	rt.keyNodes.Store(map[string]address.NodeID{})
+	return rt
+}
+
+func (rt *TrieRT) Self() key.KadKey {
+	return rt.self
+}
+
+// AddPeer tries to add a peer to the routing table
+func (rt *TrieRT) AddPeer(ctx context.Context, node address.NodeID) (bool, error) {
+	kk := node.Key()
+	if kk.Size() != rt.self.Size() {
+		return false, ErrKeyWrongLength
+	}
+
+	rt.keymu.Lock()
+	defer rt.keymu.Unlock()
+	// load the old trie, derive a mutated variant and store it in place of the original
+	keys := rt.keys.Load().(*trie.Trie)
+	keysNext := trie.Add(keys, tkey.Key(kk))
+
+	// if trie is unchanged then we didn't add the key
+	if keysNext == keys {
+		return false, nil
+	}
+
+	// make a copy of keyNodes
+	// could avoid this if we held key/nodeid tuple in the trie
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+	keyNodesNext := make(map[string]address.NodeID, len(keyNodes)+1)
+	for k, v := range keyNodes {
+		keyNodesNext[k] = v
+	}
+	keyNodesNext[string(kk)] = node
+
+	rt.keyNodes.Store(keyNodesNext)
+	rt.keys.Store(keysNext)
+	return true, nil
+}
+
+// RemoveKey tries to remove a peer identified by its Kademlia key from the
+// routing table
+func (rt *TrieRT) RemoveKey(ctx context.Context, kk key.KadKey) (bool, error) {
+	if kk.Size() != rt.self.Size() {
+		return false, ErrKeyWrongLength
+	}
+	rt.keymu.Lock()
+	defer rt.keymu.Unlock()
+	// load the old trie, derive a mutated variant and store it in place of the original
+	keys := rt.keys.Load().(*trie.Trie)
+	keysNext := trie.Remove(keys, tkey.Key(kk))
+
+	// if trie is unchanged then we didn't remove the key
+	if keysNext == keys {
+		return false, nil
+	}
+
+	// make a copy of keyNodes without removed key
+	// could avoid this if we held key/nodeid tuple in the trie
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+	if _, exists := keyNodes[string(kk)]; exists {
+		keyNodesNext := make(map[string]address.NodeID, len(keyNodes))
+		for k, v := range keyNodes {
+			if k == string(kk) {
+				continue
+			}
+			keyNodesNext[k] = v
+		}
+		rt.keyNodes.Store(keyNodesNext)
+	}
+
+	rt.keys.Store(keysNext)
+	return true, nil
+}
+
+// NearestPeers returns the n closest peers to a given key
+func (rt *TrieRT) NearestPeers(ctx context.Context, kk key.KadKey, n int) ([]address.NodeID, error) {
+	if kk.Size() != rt.self.Size() {
+		return nil, ErrKeyWrongLength
+	}
+
+	keys := rt.keys.Load().(*trie.Trie)
+	closestKeys, err := closestAtDepth(ctx, tkey.Key(kk), keys, 0, n, make([]tkey.Key, 0, n))
+	if err != nil {
+		return nil, err
+	}
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+
+	nodes := make([]address.NodeID, 0, len(closestKeys))
+	for _, c := range closestKeys {
+		if id, ok := keyNodes[string(key.KadKey(c))]; ok {
+			nodes = append(nodes, id)
+		}
+	}
+
+	return nodes, nil
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
+	if ctx.Err() != nil {
```

Comment:
is `ctx` ever expected to be cancelled during an operation?

Reply:
Generally I like to allow cancellation on long loops to improve responsiveness, but since this is doing very little work in each iteration then I agree it's not needed.